rest-api/src/main/java/com/cloudbees/workflow/flownode/FlowNodeUtil.java

```diff
@@ -267,36 +267,33 @@ public static WorkflowRun getWorkflowRunForExecution(@CheckForNull FlowExecution
     @Extension
     public static class RenameHandler extends ItemListener {
 
-        @Override
-        public void onLocationChanged(Item item, String oldFullName, String newFullName) {
-            // Better solution: scam through cache, find all instances where oldFullName matches
-            // Replace with newFullName
+        /** Removes all cache entries, because the pipeline has been deleted/renamed.
+         *  Works by scanning the cache -- which is faster than iterating all builds because
+         *  it does not require deserializing build records, and cache is capped at 1000 entries.
+         */
+        private void removeCachedRuns(String pipelineFullName) {
+            String runPrefix = pipelineFullName+"#"; // See Run#getExternalizableId - this is hardcoded
             CacheExtension ext = CacheExtension.all().get(0);
             Cache<String, RunExt> rc = ext.getRunCache();
-
-            if (item instanceof WorkflowJob) {
-                RunList<WorkflowRun> runs = ((WorkflowJob) item).getBuilds().limit(JobExt.MAX_RUNS_PER_JOB+5);  // Add a few to help invalidate just-completed
-                for (WorkflowRun r : runs) {
-                    if (!r.isBuilding()) {
-                        String path = oldFullName+"#"+r.getId();
-                        RunExt cachedRun = rc.getIfPresent(path);
-                        if (cachedRun != null) {
-                            rc.invalidate(path);
-                            rc.put(newFullName+"#"+r.getId(), cachedRun);
-                        }
-                    }
+            ConcurrentMap<String, RunExt> runMap = rc.asMap();
+            for (String cacheRunId : runMap.keySet()) {  // Put the Concurrent in ConcurrentMap to work for us
+                // Null-check may not be needed, but just in case of mutation by another thread
```

Comment:
@jglick  I hope not, but am feeling a bit cautious - plus it's easier to put the check than to prove it safe.